pkg/store/storepb/matcher_cache.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storepb
+
+import (
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const DefaultCacheSize = 200
+
+type NewItemFunc func(matcher LabelMatcher) (*labels.Matcher, error)
+
+type MatchersCache struct {
+	reg     prometheus.Registerer
+	cache   *lru.TwoQueueCache[LabelMatcher, *labels.Matcher]
+	metrics *matcherCacheMetrics
+	size    int
+}
+
+type MatcherCacheOption func(*MatchersCache)
+
+func WithPromRegistry(reg prometheus.Registerer) MatcherCacheOption {
+	return func(c *MatchersCache) {
+		c.reg = reg
+	}
+}
+
+func WithSize(size int) MatcherCacheOption {
+	return func(c *MatchersCache) {
+		c.size = size
+	}
+}
+
+func NewMatchersCache(opts ...MatcherCacheOption) (*MatchersCache, error) {
```

Comment:
Maybe we can just use `pkg/cache/inmemory.go`? It's another LRU implementation that already exists in the tree.

Reply:
I would need to make it generic first, no? Or you mean adding this LRU to also be stored there? Also, I feel like this would introduce the need for the user to configure the cache via YAML configuration in the receiver for example, which would get quite complex.